### DISCUSSION TITLE
[ZK Stack] - initcode too large

### DISCUSSION THREAD

**Discussion by sachanayush47** [zkSync-Community-Hub/zksync-developers#714] [ZK Stack] [answered]
### Module Affected

- [X] System contracts (era-system-contracts)
- [X] Smart contracts (era-contracts)
- [ ] zkEVM (era-zk_evm)
- [ ] Server (zksync-era)
- [ ] Prover (era-boojum)
- [ ] Fee withdrawer (era-fee-withdrawer)
- [ ] Block Explorer (clock-explorer)
- [ ] Portal - Wallet + Bridge (dapp-portal)
- [ ] Other (please specify below)

### Rust Version

1.80.1

### Operating System

- [X] Ubuntu
- [ ] MacOS
- [ ] Windows
- [ ] Other (Please specify below)

### Issue Description

I was trying to deploy Zk stack on holesky and I encountered this error:
Im using Alchemy's growth plan. Please check the below image.

Edit: Tutorial Im following:  https://code.zksync.io/tutorials/custom-zk-chain

![image_2024-10-04_15-49-03](https://github.com/user-attachments/assets/fbedf025-4d39-455d-a766-399cae224f83)

Thanks

### Expected Behavior

Contracts get deployed

### Current Behavior

Contracts not getting deployed.

### Prior Research

- [X] Yes, I searched for similar issues and did not find any matching ones.

**Comment by alhenaai** [comment]
It appears you are encountering an issue with the `initcode` size being too large when trying to deploy ZK Stack on the Holesky testnet. This is a common problem when the bytecode generated by the compilation of smart contracts exceeds the limits that the network can handle.

To address this issue, you can try the following solution:

- **Optimize Bytecode Size**: Modify your `hardhat.config.ts` file to include the `mode: 3` flag, which optimizes the bytecode size during compilation. This can help reduce the size of the generated bytecode, making it possible to deploy your contracts within the size limits of the network.

For more detailed guidance on optimizing your contract's bytecode, you can refer to the [hardhat-zksync-solc configuration](https://docs.zksync.io/build/tooling/hardhat/plugins/hardhat-zksync-solc#configuration).

If you continue to face issues, please ensure that your contracts do not inherently exceed the size limits due to their complexity. In such cases, consider simplifying the contracts or splitting them into smaller modules.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by iscxyz** [comment]
Hello, thanks for the report. 

To confirm my understanding - are you running the `zk_toolbox` and selecting Holesky as the L1 chain? `zk_toolbox` is using `foundry-rs` to deploy the L2 contracts to Holesky. I'd recommend ensuring that `foundry-rs` is properly installed. 

Have you tried deploying your contracts to a local node? Does that work?

**Reply by sachanayush47** [reply]
> Hello, thanks for the report.
> 
> To confirm my understanding - are you running the `zk_toolbox` and selecting Holesky as the L1 chain? `zk_toolbox` is using `foundry-rs` to deploy the L2 contracts to Holesky. I'd recommend ensuring that `foundry-rs` is properly installed.
> 
> Have you tried deploying your contracts to a local node? Does that work?

I'm following this tutorial https://code.zksync.io/tutorials/custom-zk-chain to deploy Zk stack. This tutorial doesnt involves zk_toolbox.

**Reply by bxpana** [reply]
The `zk_inception` cli is part of the zk_toolbox. When you do `forge --version` what do you get? You should get `0.2.0` if not then you might be running the ZKsync version of Foundry and need to switch by running `foundryup`

**Reply by sachanayush47** [reply]
> The `zk_inception` cli is part of the zk_toolbox. When you do `forge --version` what do you get? You should get `0.2.0` if not then you might be running the ZKsync version of Foundry and need to switch by running `foundryup`

Output of `forge --version`
`forge 0.2.0 (d90e997d9 2024-09-03T12:34:29.216253797Z)`

**Reply by dutterbutter** [reply]
Let me see if I can reproduce and work through it.

**Reply by dutterbutter** [reply]
Hey @sachanayush47 I wasnt able to reproduce and deployed / registered successfully using Holesky. Can you please provide more details on your steps? 

Given the screenshot it looks as if you are running the forge script yourself rather then utilizing `zk_inception` ?

**Reply by sachanayush47** [reply]
> Hey @sachanayush47 I wasnt able to reproduce and deployed / registered successfully using Holesky. Can you please provide more details on your steps?
> 
> Given the screenshot it looks as if you are running the forge script yourself rather then utilizing `zk_inception` ?

Hi @dutterbutter Im using zk_inception to deploy on holesky and Im not running forge script directly. By the way, which RPC provider you were using for deployment?

Below are the exact commands Im running to deploy on holesky.

Command 1: Creating ecosystem, using Validium with custom base token and mock gpu prover.
`zk_inception ecosystem create --ecosystem-name ecosystem_holesky --l1-network holesky --chain-name validium_holesky --chain-id 271 --prover-mode gpu --wallet-creation empty --l1-batch-commit-data-generator-mode validium --base-token-address 0xf2D41a7642d6F62BB3F927e1bA86F0b144ae92AA --base-token-price-nominator 20 --base-token-price-denominator 1 --start-containers true`

Command 2: Initializing ecosystem and chain.
`zk_inception ecosystem init -v -u --deploy-paymaster true --l1-rpc-url https://eth-holesky.g.alchemy.com/v2/W-FEBNPHBcJk5aHyLtImZGExpeyGQIOf --deploy-ecosystem true --deploy-erc20 false -o`

You can try using the RPC from comand 2 to reproduce.
Error is thrown by the command 2 while initializing the chain. There are  no errors while initializing the ecosystem.

Thanks

**Reply by dutterbutter** [reply]
Thanks for the detailed steps! I was using Infura paid service but seems I hit my daily requests with my reproduction here but I was able to reproduce! So let me investigate further.

**Reply by dutterbutter** [reply]
@sachanayush47 for the base token (e.g. `0xf2D41a7642d6F62BB3F927e1bA86F0b144ae92AA`) did you follow the basic ERC20 example provided in the [tutorial](https://code.zksync.io/tutorials/custom-zk-chain/customizing-your-chain#deploying-an-erc20-token)? Or is it more custom?

**Reply by sachanayush47** [reply]
> @sachanayush47 for the base token (e.g. `0xf2D41a7642d6F62BB3F927e1bA86F0b144ae92AA`) did you follow the basic ERC20 example provided in the [tutorial](https://code.zksync.io/tutorials/custom-zk-chain/customizing-your-chain#deploying-an-erc20-token)? Or is it more custom?

@dutterbutter  It is same as example in the tutorial.

**Reply by dutterbutter** [reply, marked as answer]
It looks like others have intermittently experienced a similar issue and was resolved using a holsky eth rpc with bigger limits and creating new docker containers. I can confirm using Ankr paid plan. Please see: https://github.com/matter-labs/zksync-era/issues/2726.The issue is isolated to Holesky RPC rather then zk_inception.
